pkg/flow/internal/controller/component_node.go

```diff
@@ -0,0 +1,34 @@
+package controller
+
+import (
+	"github.com/grafana/agent/component"
+	"github.com/grafana/agent/pkg/flow/internal/dag"
+)
+
+type ComponentNode interface {
```

Comment:
As a nitpick, I'm personally not a fan of breaking up functionality into different files; it makes browsing and refactoring the codebase harder.

But could we make some more effort on here? I'm not sure why this is an extension interface, what it's use is, or what to expect from implementations.

Reply:
Added docs and improved naming. The interfaces are used to give specific roles to nodes so that they can be handled as groups. The current setup feels still far from perfect but I'm struggling to find something better without increasing the blast radius which is already important with this PR. I'm open to suggestions